docs/src/adr/018adr-inlining.md

```diff
@@ -0,0 +1,156 @@
+```tla
+GlobalA(p, q) ==
+  LET LocalB(r) == r * r
+  IN LocalB(p + q)
+```
+
+defines a global operator `GlobalA`, within which there is a locally defined `LocalB`.
+
+Suppose we are given an invariant `GlobalA(1,2) = 9`. How do we evaluate whether or not this invariant holds? To do that, we need to evaluate 
+`LET LocalB(r) == r * r IN LocalB(p + q)`, and to do _that_, we need to evaluate `LocalB(p + q)`. 
+However, we cannot evaluate `LocalB(p + q)` in a vacuum, because `p` and `q` are not values we can reason about, but instead formal parameters. 
+What we need to do, is determine the value of "`LocalB(p + q)`, if `p = 1` and `q = 2`". In other words, we need to apply the substitution `{p -> 1, q ->2}` to `LocalB(p + q)`, which gives us `LocalB(1 + 2)`. 
+Repeating this process, we apply the substitution `{r -> 1 + 2}` to 
+`r * r`, the body of `LocalB`, to obtain the following equivalence:
+```tla
+GlobalA(1,2) = 9 <=> (1 + 2) * (1 + 2) = 9
+```
+
+The process of applying these substitutions as syntactic transformations is called inlining. 
+
+More precisely, suppose we are given a _non-recursive_ operator `A` with the following definition:
+```tla
+A(p1,...,pn) == body
+```
+
+The term "inlining" (of `A`) typically refers to the process of replacing instances of operator application `A(e1,...,en)` with `body[e1/p1,...,en/pn]`, i.e. the expression obtained by replacing each instance of `pi` with `ei` within `body`.
+We elect to use the term in a broader sense of "replacing an operator with its definition", and define two "flavors" of inlining:
+  1. Standard inlining: the instantiation described above
+      1. Non-nullary inlining: the instantiation described above, except the inlining skips nullary LET-defined operators
+  2. Pass-by-name inlining: replacing an operator name `A` used as an argument to a higher-order operator with a local LET-definition: `LET A_LOCAL(p1,...,pn) == A(p1,...,pn) IN A_LOCAL`
+
+The reason for doing (1) is that, at some point, a rewriting rule would have to generate constraints from `A(e1,...,en)`. 
+To do this, we couldn't just separately encode `body` and `e1,...,en`, because the richness of the data structures allowed in TLA+ makes it difficult to combine independently generated constraints, in cases where the operator parameters are complex expressions (e.g. `e1` is some record with varied nested constructs).
+This is mostly due to the fact that there is no 1-to-1 correspondence between TLA+ objects and SMT datatypes, so encoding object equality is more complicated (which would be needed to express that `ei` instantiates `pi`).
+Therefore we must, no later than at the point of the rewriting rule, know `body[e1/p1,...,en/pn]`. 
+
+While inlining non-nullary operators is  strictly necessary, inlining nullary operators is not, because nullary operators, by definition, do not have formal parameters. 
```

Comment:
```suggestion
While inlining non-nullary operators is strictly necessary, inlining nullary operators is not, because nullary operators, by definition, do not have formal parameters. 
```